File: packages/agent-portal-sdk/src/gateway-runtime-client/gateway-runtime-sandbox-operations.ts
import {
	GatewayRuntimeTrustedInvocationContextSchema,
	SandboxEnvironmentCloseResultSchema,
	SandboxEnvironmentHandleRequestSchema,
	SandboxEnvironmentOpenRequestSchema,
	SandboxEnvironmentOpenResultSchema,
	SandboxEnvironmentStatusResultSchema,
	SandboxExecCancelRequestSchema,
	SandboxExecCancelResultSchema,
	SandboxExecStartRequestSchema,
	SandboxExecStartResultSchema,
	SandboxExecWaitRequestSchema,
	SandboxExecWaitResultSchema,
	SandboxFsListRequestSchema,
	SandboxFsListResultSchema,
	SandboxFsMkdirRequestSchema,
	SandboxFsMkdirResultSchema,
	SandboxFsReadRequestSchema,
	SandboxFsReadResultSchema,
	SandboxFsRemoveRequestSchema,
	SandboxFsRemoveResultSchema,
	SandboxFsRenameRequestSchema,
	SandboxFsRenameResultSchema,
	SandboxFsStatRequestSchema,
	SandboxFsStatResultSchema,
	SandboxFsWriteRequestSchema,
	SandboxFsWriteResultSchema,
	SandboxProcessCancelRequestSchema,
	SandboxProcessCancelResultSchema,
	SandboxProcessHandleRequestSchema,
	SandboxProcessLogsRequestSchema,
	SandboxProcessLogsResultSchema,
	SandboxProcessStartRequestSchema,
	SandboxProcessStartResultSchema,
	SandboxProcessStatusResultSchema,
	SandboxProcessWaitRequestSchema,
	SandboxProcessWaitResultSchema,
	SandboxRetainedResultLookupRequestSchema,
	SandboxRetainedResultLookupResultSchema,
	SandboxStreamCloseRequestSchema,
	SandboxStreamCloseResultSchema,
	SandboxStreamReadRequestSchema,
	SandboxStreamReadResultSchema,
	SandboxStreamWriteRequestSchema,
	SandboxStreamWriteResultSchema,
	SandboxTerminalAttachRequestSchema,
	SandboxTerminalAttachResultSchema,
	SandboxTerminalResizeRequestSchema,
	SandboxTerminalResizeResultSchema,
	type GatewayRuntimeTrustedInvocationContext,
	type SandboxEnvironmentCloseRequest,
	type SandboxEnvironmentCloseResult,
	type SandboxEnvironmentOpenRequest,
	type SandboxEnvironmentOpenResult,
	type SandboxEnvironmentStatusRequest,
	type SandboxEnvironmentStatusResult,
	type SandboxExecCancelRequest,
	type SandboxExecCancelResult,
	type SandboxExecStartRequest,
	type SandboxExecStartResult,
	type SandboxExecWaitRequest,
	type SandboxExecWaitResult,
	type SandboxFsListRequest,
	type SandboxFsListResult,
	type SandboxFsMkdirRequest,
	type SandboxFsMkdirResult,
	type SandboxFsReadRequest,
	type SandboxFsReadResult,
	type SandboxFsRemoveRequest,
	type SandboxFsRemoveResult,
	type SandboxFsRenameRequest,
	type SandboxFsRenameResult,
	type SandboxFsStatRequest,
	type SandboxFsStatResult,
	type SandboxFsWriteRequest,
	type SandboxFsWriteResult,
	type SandboxProcessCancelRequest,
	type SandboxProcessCancelResult,
	type SandboxProcessLogsRequest,
	type SandboxProcessLogsResult,
	type SandboxProcessStartRequest,
	type SandboxProcessStartResult,
	type SandboxProcessStatusRequest,
	type SandboxProcessStatusResult,
	type SandboxProcessWaitRequest,
	type SandboxProcessWaitResult,
	type SandboxRetainedResultLookupRequest,
	type SandboxRetainedResultLookupResult,
	type SandboxStreamCloseRequest,
	type SandboxStreamCloseResult,
	type SandboxStreamReadRequest,
	type SandboxStreamReadResult,
	type SandboxStreamWriteRequest,
	type SandboxStreamWriteResult,
	type SandboxTerminalAttachRequest,
	type SandboxTerminalAttachResult,
	type SandboxTerminalResizeRequest,
	type SandboxTerminalResizeResult,
} from '../contracts/index.js';
import type { GatewayRuntimeTraceContext } from './gateway-runtime-trace-context.js';

export interface GatewayRuntimeTrustedRequestOptions {
	readonly signal?: AbortSignal;
	readonly traceContext?: GatewayRuntimeTraceContext;
	readonly trustedContext: GatewayRuntimeTrustedInvocationContext;
}

interface GatewayRuntimeSandboxRequestClient {
	readonly request: (
		method: string,
		params: unknown,
		options: {
			readonly signal?: AbortSignal;
			readonly traceContext?: GatewayRuntimeTraceContext;
		},
	) => Promise<unknown>;
}

interface GatewayRuntimeCanonicalSchema<TValue> {
	readonly parse: (value: unknown) => TValue;
}

interface SandboxOperationRequestProps<TRequest, TResult> {
	readonly client: GatewayRuntimeSandboxRequestClient;
	readonly method: string;
	readonly options: GatewayRuntimeTrustedRequestOptions;
	readonly request: TRequest;
	readonly requestSchema: GatewayRuntimeCanonicalSchema<TRequest>;
	readonly resultSchema: GatewayRuntimeCanonicalSchema<TResult>;
}

async function requestSandboxOperation<TRequest, TResult>(
	props: SandboxOperationRequestProps<TRequest, TResult>,
): Promise<TResult> {
	const publicRequest = props.requestSchema.parse(props.request);
	const trustedContext = GatewayRuntimeTrustedInvocationContextSchema.parse(
		props.options.trustedContext,
	);
	const result = await props.client.request(
		props.method,
		{ publicRequest, trustedContext },
		{
			...(props.options.signal === undefined ? {} : { signal: props.options.signal }),
			...(props.options.traceContext === undefined
				? {}
				: { traceContext: props.options.traceContext }),
		},
	);
	return props.resultSchema.parse(result);
}

export class GatewayRuntimeSandboxEnvironmentOperations {
	readonly #client: GatewayRuntimeSandboxRequestClient;

	constructor(client: GatewayRuntimeSandboxRequestClient) {
		this.#client = client;
	}

	async open(
		request: SandboxEnvironmentOpenRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxEnvironmentOpenResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.environment.open',
			options,
			request,
			requestSchema: SandboxEnvironmentOpenRequestSchema,
			resultSchema: SandboxEnvironmentOpenResultSchema,
		});
	}

	async close(
		request: SandboxEnvironmentCloseRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxEnvironmentCloseResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.environment.close',
			options,
			request,
			requestSchema: SandboxEnvironmentHandleRequestSchema,
			resultSchema: SandboxEnvironmentCloseResultSchema,
		});
	}

	async status(
		request: SandboxEnvironmentStatusRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxEnvironmentStatusResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.environment.status',
			options,
			request,
			requestSchema: SandboxEnvironmentHandleRequestSchema,
			resultSchema: SandboxEnvironmentStatusResultSchema,
		});
	}
}

export class GatewayRuntimeSandboxExecutionOperations {
	readonly #client: GatewayRuntimeSandboxRequestClient;

	constructor(client: GatewayRuntimeSandboxRequestClient) {
		this.#client = client;
	}

	async start(
		request: SandboxExecStartRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxExecStartResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.exec.start',
			options,
			request,
			requestSchema: SandboxExecStartRequestSchema,
			resultSchema: SandboxExecStartResultSchema,
		});
	}

	async wait(
		request: SandboxExecWaitRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxExecWaitResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.exec.wait',
			options,
			request,
			requestSchema: SandboxExecWaitRequestSchema,
			resultSchema: SandboxExecWaitResultSchema,
		});
	}

	async cancel(
		request: SandboxExecCancelRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxExecCancelResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.exec.cancel',
			options,
			request,
			requestSchema: SandboxExecCancelRequestSchema,
			resultSchema: SandboxExecCancelResultSchema,
		});
	}
}

export class GatewayRuntimeSandboxRetainedResultOperations {
	readonly #client: GatewayRuntimeSandboxRequestClient;

	constructor(client: GatewayRuntimeSandboxRequestClient) {
		this.#client = client;
	}

	async lookup(
		request: SandboxRetainedResultLookupRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxRetainedResultLookupResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.retained-result.lookup',
			options,
			request,
			requestSchema: SandboxRetainedResultLookupRequestSchema,
			resultSchema: SandboxRetainedResultLookupResultSchema,
		});
	}
}

export class GatewayRuntimeSandboxFilesystemOperations {
	readonly #client: GatewayRuntimeSandboxRequestClient;

	constructor(client: GatewayRuntimeSandboxRequestClient) {
		this.#client = client;
	}

	async stat(
		request: SandboxFsStatRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxFsStatResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.fs.stat',
			options,
			request,
			requestSchema: SandboxFsStatRequestSchema,
			resultSchema: SandboxFsStatResultSchema,
		});
	}

	async list(
		request: SandboxFsListRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxFsListResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.fs.list',
			options,
			request,
			requestSchema: SandboxFsListRequestSchema,
			resultSchema: SandboxFsListResultSchema,
		});
	}

	async read(
		request: SandboxFsReadRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxFsReadResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.fs.read',
			options,
			request,
			requestSchema: SandboxFsReadRequestSchema,
			resultSchema: SandboxFsReadResultSchema,
		});
	}

	async write(
		request: SandboxFsWriteRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxFsWriteResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.fs.write',
			options,
			request,
			requestSchema: SandboxFsWriteRequestSchema,
			resultSchema: SandboxFsWriteResultSchema,
		});
	}

	async mkdir(
		request: SandboxFsMkdirRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxFsMkdirResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.fs.mkdir',
			options,
			request,
			requestSchema: SandboxFsMkdirRequestSchema,
			resultSchema: SandboxFsMkdirResultSchema,
		});
	}

	async rename(
		request: SandboxFsRenameRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxFsRenameResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.fs.rename',
			options,
			request,
			requestSchema: SandboxFsRenameRequestSchema,
			resultSchema: SandboxFsRenameResultSchema,
		});
	}

	async remove(
		request: SandboxFsRemoveRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxFsRemoveResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.fs.remove',
			options,
			request,
			requestSchema: SandboxFsRemoveRequestSchema,
			resultSchema: SandboxFsRemoveResultSchema,
		});
	}
}

export class GatewayRuntimeSandboxProcessOperations {
	readonly #client: GatewayRuntimeSandboxRequestClient;

	constructor(client: GatewayRuntimeSandboxRequestClient) {
		this.#client = client;
	}

	async start(
		request: SandboxProcessStartRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxProcessStartResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.process.start',
			options,
			request,
			requestSchema: SandboxProcessStartRequestSchema,
			resultSchema: SandboxProcessStartResultSchema,
		});
	}

	async status(
		request: SandboxProcessStatusRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxProcessStatusResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.process.status',
			options,
			request,
			requestSchema: SandboxProcessHandleRequestSchema,
			resultSchema: SandboxProcessStatusResultSchema,
		});
	}

	async wait(
		request: SandboxProcessWaitRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxProcessWaitResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.process.wait',
			options,
			request,
			requestSchema: SandboxProcessWaitRequestSchema,
			resultSchema: SandboxProcessWaitResultSchema,
		});
	}

	async logs(
		request: SandboxProcessLogsRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxProcessLogsResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.process.logs',
			options,
			request,
			requestSchema: SandboxProcessLogsRequestSchema,
			resultSchema: SandboxProcessLogsResultSchema,
		});
	}

	async cancel(
		request: SandboxProcessCancelRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxProcessCancelResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.process.cancel',
			options,
			request,
			requestSchema: SandboxProcessCancelRequestSchema,
			resultSchema: SandboxProcessCancelResultSchema,
		});
	}
}

export class GatewayRuntimeSandboxStreamOperations {
	readonly #client: GatewayRuntimeSandboxRequestClient;

	constructor(client: GatewayRuntimeSandboxRequestClient) {
		this.#client = client;
	}

	async read(
		request: SandboxStreamReadRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxStreamReadResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.stream.read',
			options,
			request,
			requestSchema: SandboxStreamReadRequestSchema,
			resultSchema: SandboxStreamReadResultSchema,
		});
	}

	async write(
		request: SandboxStreamWriteRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxStreamWriteResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.stream.write',
			options,
			request,
			requestSchema: SandboxStreamWriteRequestSchema,
			resultSchema: SandboxStreamWriteResultSchema,
		});
	}

	async close(
		request: SandboxStreamCloseRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxStreamCloseResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.stream.close',
			options,
			request,
			requestSchema: SandboxStreamCloseRequestSchema,
			resultSchema: SandboxStreamCloseResultSchema,
		});
	}
}

export class GatewayRuntimeSandboxTerminalOperations {
	readonly #client: GatewayRuntimeSandboxRequestClient;

	constructor(client: GatewayRuntimeSandboxRequestClient) {
		this.#client = client;
	}

	async attach(
		request: SandboxTerminalAttachRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxTerminalAttachResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.terminal.attach',
			options,
			request,
			requestSchema: SandboxTerminalAttachRequestSchema,
			resultSchema: SandboxTerminalAttachResultSchema,
		});
	}

	async resize(
		request: SandboxTerminalResizeRequest,
		options: GatewayRuntimeTrustedRequestOptions,
	): Promise<SandboxTerminalResizeResult> {
		return await requestSandboxOperation({
			client: this.#client,
			method: 'sandbox.terminal.resize',
			options,
			request,
			requestSchema: SandboxTerminalResizeRequestSchema,
			resultSchema: SandboxTerminalResizeResultSchema,
		});
	}
}

export class GatewayRuntimeSandboxOperations {
	readonly environment: GatewayRuntimeSandboxEnvironmentOperations;
	readonly execution: GatewayRuntimeSandboxExecutionOperations;
	readonly filesystem: GatewayRuntimeSandboxFilesystemOperations;
	readonly process: GatewayRuntimeSandboxProcessOperations;
	readonly retainedResults: GatewayRuntimeSandboxRetainedResultOperations;
	readonly stream: GatewayRuntimeSandboxStreamOperations;
	readonly terminal: GatewayRuntimeSandboxTerminalOperations;

	constructor(client: GatewayRuntimeSandboxRequestClient) {
		this.environment = new GatewayRuntimeSandboxEnvironmentOperations(client);
		this.execution = new GatewayRuntimeSandboxExecutionOperations(client);
		this.filesystem = new GatewayRuntimeSandboxFilesystemOperations(client);
		this.process = new GatewayRuntimeSandboxProcessOperations(client);
		this.retainedResults = new GatewayRuntimeSandboxRetainedResultOperations(client);
		this.stream = new GatewayRuntimeSandboxStreamOperations(client);
		this.terminal = new GatewayRuntimeSandboxTerminalOperations(client);
		Object.freeze(this.environment);
		Object.freeze(this.execution);
		Object.freeze(this.filesystem);
		Object.freeze(this.process);
		Object.freeze(this.retainedResults);
		Object.freeze(this.stream);
		Object.freeze(this.terminal);
	}
}

File: python/agent-vm-agent-portal-sdk/src/agent_vm_agent_portal_sdk/gateway_runtime_sandbox_operations.py
"""Portable rich-sandbox operations for one Gateway Runtime attachment."""

import typing as t
from collections.abc import Mapping

from pydantic import BaseModel

from .contracts import PORTABLE_CONTRACT_ADAPTERS

type SandboxWireMethod = t.Literal[
    "sandbox.environment.open",
    "sandbox.environment.close",
    "sandbox.environment.status",
    "sandbox.exec.start",
    "sandbox.exec.wait",
    "sandbox.exec.cancel",
    "sandbox.retained-result.lookup",
    "sandbox.fs.stat",
    "sandbox.fs.list",
    "sandbox.fs.read",
    "sandbox.fs.write",
    "sandbox.fs.mkdir",
    "sandbox.fs.rename",
    "sandbox.fs.remove",
    "sandbox.process.start",
    "sandbox.process.status",
    "sandbox.process.wait",
    "sandbox.process.logs",
    "sandbox.process.cancel",
    "sandbox.stream.read",
    "sandbox.stream.write",
    "sandbox.stream.close",
    "sandbox.terminal.attach",
    "sandbox.terminal.resize",
]


SANDBOX_WIRE_METHOD_SCHEMA_IDS: t.Final[dict[SandboxWireMethod, tuple[str, str]]] = {
    "sandbox.retained-result.lookup": (
        "sandbox.retained-result.lookup-request",
        "sandbox.retained-result.lookup-result",
    ),
}


def _sandbox_method_schema_ids(method: SandboxWireMethod) -> tuple[str, str]:
    explicit_schema_ids = SANDBOX_WIRE_METHOD_SCHEMA_IDS.get(method)
    if explicit_schema_ids is not None:
        return explicit_schema_ids
    return f"{method}.request", f"{method}.result"


class GatewayRuntimeSandboxRequestClient(t.Protocol):
    async def request(
        self,
        method: str,
        params: Mapping[str, object],
    ) -> Mapping[str, object]: ...


def _validate_portable_mapping(
    schema_id: str,
    value: Mapping[str, object],
) -> dict[str, object]:
    validated_value = PORTABLE_CONTRACT_ADAPTERS[schema_id].validate_python(dict(value))
    if not isinstance(validated_value, BaseModel):
        invalid_mapping_message = f"Portable contract {schema_id!r} did not produce a typed model."
        raise TypeError(invalid_mapping_message)
    normalized_value = validated_value.model_dump(
        by_alias=True,
        exclude_none=True,
        mode="json",
    )
    if not isinstance(normalized_value, dict):
        invalid_mapping_message = f"Portable contract {schema_id!r} did not produce a JSON object."
        raise TypeError(invalid_mapping_message)
    return t.cast("dict[str, object]", normalized_value)


async def _execute_sandbox_operation(
    *,
    client: GatewayRuntimeSandboxRequestClient,
    method: SandboxWireMethod,
    public_request: Mapping[str, object],
    trusted_context: Mapping[str, object],
) -> BaseModel:
    request_schema_id, result_schema_id = _sandbox_method_schema_ids(method)
    normalized_public_request = _validate_portable_mapping(
        request_schema_id,
        public_request,
    )
    normalized_trusted_context = _validate_portable_mapping(
        "gateway.trusted-invocation-context",
        trusted_context,
    )
    response = await client.request(
        method,
        {
            "publicRequest": normalized_public_request,
            "trustedContext": normalized_trusted_context,
        },
    )
    validated_result = PORTABLE_CONTRACT_ADAPTERS[result_schema_id].validate_python(dict(response))
    if not isinstance(validated_result, BaseModel):
        invalid_result_message = f"Sandbox operation {method!r} did not produce a typed result."
        raise TypeError(invalid_result_message)
    return validated_result


class _GatewayRuntimeSandboxOperationGroup:
    def __init__(self, client: GatewayRuntimeSandboxRequestClient) -> None:
        self._client = client

    async def _execute(
        self,
        method: SandboxWireMethod,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await _execute_sandbox_operation(
            client=self._client,
            method=method,
            public_request=request,
            trusted_context=trusted_context,
        )


class GatewayRuntimeSandboxEnvironmentOperations(_GatewayRuntimeSandboxOperationGroup):
    async def open(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute(
            "sandbox.environment.open",
            request,
            trusted_context=trusted_context,
        )

    async def close(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute(
            "sandbox.environment.close",
            request,
            trusted_context=trusted_context,
        )

    async def status(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute(
            "sandbox.environment.status",
            request,
            trusted_context=trusted_context,
        )


class GatewayRuntimeSandboxExecutionOperations(_GatewayRuntimeSandboxOperationGroup):
    async def start(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute(
            "sandbox.exec.start",
            request,
            trusted_context=trusted_context,
        )

    async def wait(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute(
            "sandbox.exec.wait",
            request,
            trusted_context=trusted_context,
        )

    async def cancel(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute(
            "sandbox.exec.cancel",
            request,
            trusted_context=trusted_context,
        )


class GatewayRuntimeSandboxRetainedResultOperations(_GatewayRuntimeSandboxOperationGroup):
    async def lookup(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute(
            "sandbox.retained-result.lookup",
            request,
            trusted_context=trusted_context,
        )


class GatewayRuntimeSandboxFilesystemOperations(_GatewayRuntimeSandboxOperationGroup):
    async def stat(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.fs.stat", request, trusted_context=trusted_context)

    async def list(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.fs.list", request, trusted_context=trusted_context)

    async def read(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.fs.read", request, trusted_context=trusted_context)

    async def write(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.fs.write", request, trusted_context=trusted_context)

    async def mkdir(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.fs.mkdir", request, trusted_context=trusted_context)

    async def rename(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.fs.rename", request, trusted_context=trusted_context)

    async def remove(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.fs.remove", request, trusted_context=trusted_context)


class GatewayRuntimeSandboxProcessOperations(_GatewayRuntimeSandboxOperationGroup):
    async def start(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.process.start", request, trusted_context=trusted_context)

    async def status(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.process.status", request, trusted_context=trusted_context)

    async def wait(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.process.wait", request, trusted_context=trusted_context)

    async def logs(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.process.logs", request, trusted_context=trusted_context)

    async def cancel(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.process.cancel", request, trusted_context=trusted_context)


class GatewayRuntimeSandboxStreamOperations(_GatewayRuntimeSandboxOperationGroup):
    async def read(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.stream.read", request, trusted_context=trusted_context)

    async def write(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.stream.write", request, trusted_context=trusted_context)

    async def close(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.stream.close", request, trusted_context=trusted_context)


class GatewayRuntimeSandboxTerminalOperations(_GatewayRuntimeSandboxOperationGroup):
    async def attach(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.terminal.attach", request, trusted_context=trusted_context)

    async def resize(
        self,
        request: Mapping[str, object],
        *,
        trusted_context: Mapping[str, object],
    ) -> BaseModel:
        return await self._execute("sandbox.terminal.resize", request, trusted_context=trusted_context)


class GatewayRuntimeSandboxOperations:
    """All portable rich-sandbox groups backed by one Gateway Runtime client."""

    def __init__(self, client: GatewayRuntimeSandboxRequestClient) -> None:
        self.environment: GatewayRuntimeSandboxEnvironmentOperations = GatewayRuntimeSandboxEnvironmentOperations(client)
        self.execution: GatewayRuntimeSandboxExecutionOperations = GatewayRuntimeSandboxExecutionOperations(client)
        self.filesystem: GatewayRuntimeSandboxFilesystemOperations = GatewayRuntimeSandboxFilesystemOperations(client)
        self.process: GatewayRuntimeSandboxProcessOperations = GatewayRuntimeSandboxProcessOperations(client)
        self.retained_results: GatewayRuntimeSandboxRetainedResultOperations = GatewayRuntimeSandboxRetainedResultOperations(client)
        self.stream: GatewayRuntimeSandboxStreamOperations = GatewayRuntimeSandboxStreamOperations(client)
        self.terminal: GatewayRuntimeSandboxTerminalOperations = GatewayRuntimeSandboxTerminalOperations(client)
